[kernels] Migrate comm/ from LegacyUnsafePointer to UnsafePointer

KrxGu opened the pull request:
Fixes #5673 

## Summary

Migrates the comm module and its dependents from `LegacyUnsafePointer` to the new `UnsafePointer` API with explicit mutability and origin parameters.

Part 1 of #5673.

## Changes

**comm/ module:**
- `sync.mojo`: Updated imports, replaced `LegacyOpaquePointer` with `MutOpaquePointer[MutOrigin.external]` for FFI callbacks
- `allgather.mojo`: Added explicit `mut` and `MutAnyOrigin` to all pointer signatures
- `allreduce.mojo`: Migrated all pointer usages to new API

**Dependent files:**
- `linalg/distributed_matmul.mojo`: Updated import and `rank_sigs` signatures
- `Mogg/MOGGKernelAPI/MOGGKernelAPI.mojo`: Added `RealUnsafePointer` alias for comm API calls while preserving `LegacyUnsafePointer` for other usages

## Testing

Builds successfully:
- `./bazelw build //max/kernels/src/comm:comm`
- `./bazelw build //max/kernels/src/linalg:linalg`
- `./bazelw build //max/kernels/src/Mogg/MOGGKernelAPI:MOGGKernelAPI`

## Next Steps

Follow-up PRs will migrate `shmem/` and `comm/vendor/ccl.mojo`.

KrxGu committed:
Migrate comm/ from LegacyUnsafePointer to UnsafePointer

Replace LegacyUnsafePointer with the new UnsafePointer API in sync.mojo,
allgather.mojo, and allreduce.mojo.

Key changes:
- Use explicit mut=True/False and origin parameters
- Update OpaquePointer to MutOpaquePointer for FFI callbacks
- MutAnyOrigin for pointer types that need to work across origins

Part of #5673

KrxGu committed:
Fix missing mut parameters and migrate distributed_matmul.mojo

- Add explicit mut=True/False to pointer signatures in allreduce.mojo
- Migrate distributed_matmul.mojo from LegacyUnsafePointer to UnsafePointer
- Update all rank_sigs type signatures to use new pointer API

KrxGu committed:
Fix MOGGKernelAPI.mojo to use new UnsafePointer for comm APIs

Use RealUnsafePointer alias for rank_sigs in allreduce, allgather,
and matmul_allreduce calls while keeping LegacyUnsafePointer for
existing usages in the file.

NathanSWard commented on max/kernels/src/comm/allgather.mojo:
```diff
@@ -162,14 +168,18 @@ fn _allgather_p2p[
     # Prepare input pointers
-    var list_of_in_ptrs = StaticTuple[UnsafePointer[Scalar[dtype]], ngpus]()
+    var list_of_in_ptrs = StaticTuple[
+        UnsafePointer[mut=False, Scalar[dtype], MutAnyOrigin], ngpus
```
```suggestion
        UnsafePointer[Scalar[dtype], ImmutAnyOrigin], ngpus
```

Assuming these are `immutable` pointers you should be able to use `ImmuAnyOrigin` here!

NathanSWard commented on max/kernels/src/comm/allgather.mojo:
```diff
@@ -184,7 +194,9 @@ fn _allgather_p2p[
         # Prepare output pointers for this GPU.
-        var output_ptrs = StaticTuple[UnsafePointer[Scalar[dtype]], ngpus]()
+        var output_ptrs = StaticTuple[
+            UnsafePointer[mut=True, Scalar[dtype], MutAnyOrigin], ngpus
```
```suggestion
            UnsafePointer[Scalar[dtype], MutAnyOrigin], ngpus
```
tip: If you explicitly specify the `origin` of the pointer, you don't need to explicitly specify the `mut` parameter :)

NathanSWard commented on max/kernels/src/comm/allgather.mojo:
```diff
@@ -234,7 +246,9 @@ fn allgather[
     ],
-    rank_sigs: InlineArray[UnsafePointer[Signal], MAX_GPUS],
+    rank_sigs: InlineArray[
+        UnsafePointer[mut=True, Signal, MutAnyOrigin], MAX_GPUS
```
```suggestion
        UnsafePointer[Signal, MutAnyOrigin], MAX_GPUS
```

NathanSWard commented on max/kernels/src/comm/allreduce.mojo:
```diff
@@ -153,8 +151,8 @@ comptime _target_address_space = AddressSpace.GLOBAL if is_amd_gpu() else Addres
-    dst_buf: UnsafePointer[Scalar[dtype]],
-    src_buf: UnsafePointer[Scalar[dtype]],
+    dst_buf: UnsafePointer[mut=True, Scalar[dtype], MutAnyOrigin],
+    src_buf: UnsafePointer[mut=False, Scalar[dtype], MutAnyOrigin],
```
When `UnsafePointer` is used as a function argument - you can leave the `origin` unbound, instead of forcing it to be the "any" origin.
```mojo
    dst_buf: UnsafePointer[mut=True, Scalar[dtype]],
    src_buf: UnsafePointer[mut=False, Scalar[dtype]],
```

KrxGu replied:
makes sense!!

NathanSWard commented on max/kernels/src/comm/allreduce.mojo:
```diff
@@ -187,7 +185,7 @@ fn _naive_reduce_kernel_with_lambda[
 ](
     dst_buf: NDBuffer[dtype, rank, MutAnyOrigin],
-    src_buf: UnsafePointer[Scalar[dtype]],
+    src_buf: UnsafePointer[mut=False, Scalar[dtype], MutAnyOrigin],
```
```suggestion
    src_buf: UnsafePointer[mut=False, Scalar[dtype]],
```
Ditto leaving the origin unbound.

NathanSWard commented on max/kernels/src/comm/allreduce.mojo:
```diff
@@ -348,7 +346,9 @@ fn _load_reduce[
     elem_idx: Int,
-    ptrs: InlineArray[UnsafePointer[Scalar[dtype]], num_buffers],
+    ptrs: InlineArray[
+        UnsafePointer[mut=False, Scalar[dtype], MutAnyOrigin], num_buffers
```
```suggestion
        UnsafePointer[Scalar[dtype], ImmutAnyOrigin], num_buffers
```

NathanSWard commented on max/kernels/src/comm/allreduce.mojo:
```diff
@@ -401,8 +401,12 @@ fn _allreduce_2stage_kernel[
     result: NDBuffer[dtype, rank, MutAnyOrigin],
-    src_ptrs: InlineArray[UnsafePointer[Scalar[dtype]], num_buffers],
-    rank_sigs: InlineArray[UnsafePointer[Signal], MAX_GPUS],
+    src_ptrs: InlineArray[
+        UnsafePointer[mut=False, Scalar[dtype], MutAnyOrigin], num_buffers
+    ],
```
For all of these cases, you should be able to use `ImmutAnyOrigin` if `mut=False`, and `MutAnyOrigin`, if `mut=True`.

I'll have the comment cover the rest of the `srt_ptrs` and `rank_sigs` in the other files.

KrxGu replied:
Gotcha!!

NathanSWard commented on max/kernels/src/comm/sync.mojo:
```diff
@@ -25,11 +25,7 @@ from gpu.intrinsics import (
-    LegacyUnsafePointer as UnsafePointer,
-    UnsafePointer as UnsafePointerV2,
-)
+from memory import MutOpaquePointer, UnsafePointer
```
*Note*: All of these are actually imported in the prelude, so you don't need to manually import them in the files!

NathanSWard commented on max/kernels/src/comm/sync.mojo:
```diff
@@ -41,12 +37,14 @@ fn group_end():
 
 
-fn _p2p_cache_init_wrapper() -> OpaquePointer:
+fn _p2p_cache_init_wrapper() -> MutOpaquePointer[MutOrigin.external]:
```
Nice use of `external` origin here 😄

NathanSWard reviewed (changes requested):
Thanks for tackling this!
Let a handful of comments - mostly around when needing/not needing to manually specify an `origin` or `mut` parameter.
Hopefully it makes sense, but if not please ask questions and I'll do my best to help/explain!

KrxGu committed:
Address Nathan's review feedback

- Remove unnecessary imports (types are in prelude)
- Use ImmutAnyOrigin for immutable pointers (mut=False)
- Remove mut= parameter when origin is explicit
- Keep origins explicit for GPU kernel parameters (enqueue_function_checked)
- Simplify pointer type signatures across all files

NathanSWard commented on max/kernels/src/comm/allreduce.mojo:
```diff
@@ -401,8 +401,12 @@ fn _allreduce_2stage_kernel[
     result: NDBuffer[dtype, rank, MutAnyOrigin],
-    src_ptrs: InlineArray[UnsafePointer[Scalar[dtype]], num_buffers],
-    rank_sigs: InlineArray[UnsafePointer[Signal], MAX_GPUS],
+    src_ptrs: InlineArray[
+        UnsafePointer[mut=False, Scalar[dtype], MutAnyOrigin], num_buffers
+    ],
+    rank_sigs: InlineArray[
+        UnsafePointer[mut=True, Signal, MutAnyOrigin], MAX_GPUS
+    ],
```
These can both be updated to: (dropping the `mut=` part :) 
```
src_ptrs: InlineArray[
        UnsafePointer[Scalar[dtype], ImmutAnyOrigin], num_buffers
    ],
 rank_sigs: InlineArray[
        UnsafePointer[Signal, MutAnyOrigin], MAX_GPUS
    ],

```

NathanSWard commented on max/kernels/src/comm/allreduce.mojo:
```diff
@@ -561,8 +565,12 @@ fn _allreduce_1stage_kernel[
     result: NDBuffer[dtype, rank, MutAnyOrigin],
-    src_ptrs: InlineArray[UnsafePointer[Scalar[dtype]], num_buffers],
-    rank_sigs: InlineArray[UnsafePointer[Signal], MAX_GPUS],
+    src_ptrs: InlineArray[
+        UnsafePointer[mut=False, Scalar[dtype], MutAnyOrigin], num_buffers
+    ],
+    rank_sigs: InlineArray[
+        UnsafePointer[mut=True, Signal, MutAnyOrigin], MAX_GPUS
+    ],
```
ditto here for 
```mojo
    src_ptrs: InlineArray[
        UnsafePointer[Scalar[dtype], ImmutAnyOrigin], num_buffers
    ],
    rank_sigs: InlineArray[
        UnsafePointer[Signal, MutAnyOrigin], MAX_GPUS
    ],
```

NathanSWard reviewed (commented):
After this gets rebased and the last to comments are address this looks great and I'll merge it in :D
